### PR TITLE
Sitemap REST API with long polling: observe all necessary items for …

### LINKED ISSUE #4938: Sitemap REST API with long polling: not all items are observed for update
All items used in visibility/labelcolor/valuecolor attributes which are not present in the current page are ignored, meaning the server is not listening for updates on these items.
Additionally, if you have a frame linked to an item, all widgets linked to items inside this frame are ignored too.

The consequence is that any UI using the long polling method will not be informed about the update of such items and so the page will not be updated.

The known impacted UIs are the Android app and Basic UI when used in Microsoft Edge.

I will propose a fix.

### PR DESCRIPTION
**Pull request by lolodomo**
…update

Fixes #4938

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
With again so much new logic here, it is imho really time to cover those with some tests to avoid to accidentally break this again, would you agree?

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java [hunk 1]
```diff
@@ -614,9 +616,55 @@ private boolean waitForChanges(EList<Widget> widgets) {
...
+                items.addAll(getItemsInColorCond(widget.getLabelColor()));
+                items.addAll(getItemsInColorCond(widget.getValueColor()));
+            }
+        }
+        return items;
+    }
+
+    private Set<GenericItem> getItemsInVisibilityCond(EList<VisibilityRule> ruleList) {
+        Set<GenericItem> items = new HashSet<GenericItem>();
+        if (ruleList != null) {
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java]: `rulesList` will never be null, `widget#getVisibility` always returns a list instance.

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java [hunk 2]
```diff
@@ -614,9 +616,55 @@ private boolean waitForChanges(EList<Widget> widgets) {
...
+                    }
+                }
+            }
+        }
+        return items;
+    }
+
+    private Set<GenericItem> getItemsInColorCond(EList<ColorArray> colorList) {
+        Set<GenericItem> items = new HashSet<GenericItem>();
+        if (colorList != null) {
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java]: `colorList ` will never be null, `widget#getLabelColor` and `widget#getValueColor` always return a list instance.

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java [hunk 3]
```diff
@@ -614,9 +616,55 @@ private boolean waitForChanges(EList<Widget> widgets) {
...
+            for (ColorArray color : colorList) {
+                String itemName = color.getItem();
+                if (itemName != null) {
+                    try {
+                        Item item = itemUIRegistry.getItem(itemName);
+                        if (item instanceof GenericItem) {
+                            final GenericItem gItem = (GenericItem) item;
+                            items.add(gItem);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java]: maybe matter of taste: I would prefer `gItem` to be inlined: `items.add((GenericItem) item)`.

File: bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java [hunk 4]
```diff
@@ -614,9 +616,55 @@ private boolean waitForChanges(EList<Widget> widgets) {
...
+        Set<GenericItem> items = new HashSet<GenericItem>();
+        if (ruleList != null) {
+            for (VisibilityRule rule : ruleList) {
+                String itemName = rule.getItem();
+                if (itemName != null) {
+                    try {
+                        Item item = itemUIRegistry.getItem(itemName);
+                        if (item instanceof GenericItem) {
+                            final GenericItem gItem = (GenericItem) item;
+                            items.add(gItem);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java]: maybe matter of taste: I would prefer `gItem` to be inlined: `items.add((GenericItem) item)`.

**Comment by htreu** [comment]
I second @kaikreuzer´s request for a test case.

**Comment by lolodomo** [comment]
@htreu :  I will update my PR considering your comments on code.

Regarding the tests, I am ok to add one or two test cases but I don't find where are defined the current tests for this API. I am sorry but don't count on me to create all tests stuff just because I provide a minor change (but important fix for users). i am mot enough familiar with that.

**Comment by htreu** [comment]
@lolodomo the SitemapResource is not tested at all right now. I do volunteer for creating a test which will also cover your changes. During review I already made a plan :-)

**Comment by lolodomo** [comment]
Thank you.

**Comment by htreu** [comment]
Hi @lolodomo, please the this PR against your branch: https://github.com/lolodomo/smarthome/pull/3
This adds a new test fragment and a SitemapResourceTest which already tests your new logic.

**Comment by lolodomo** [comment]
@htreu: wouldn't it be better to have your tests in a new PR ? You are adding a lot of new code.

**Comment by htreu** [comment]
@lolodomo fine for me. I will add my PR once this got merged.
@maggu2810, @sjka or @kaikreuzer your turn :-)

**Comment by kaikreuzer** [comment]
Err, why? I thought @htreu did the whole exercise so that this PR can have tests?

**Comment by lolodomo** [comment]
I merged the PR. The test bundle is building with success on my machine.

**Comment by maggu2810** [comment]
As far as I understand:
* @htreu agreed with the changes of @lolodomo
* @htreu added tests that succeeded on @lolodomo's machine

So, I assume it is good to merge. Is it?

**Comment by kaikreuzer** [comment]
Yes, now that the tests are part of the PR and succeed, we are good to merge.

**Pull request merged**